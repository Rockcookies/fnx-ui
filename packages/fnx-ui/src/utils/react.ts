import {
	Children,
	forwardRef,
	ForwardRefExoticComponent,
	ForwardRefRenderFunction,
	FunctionComponent,
	PropsWithoutRef,
	ReactElement,
	RefAttributes,
} from 'react';
import { isFragment } from 'react-is';

export const toElementArray = (
	children: React.ReactNode,
	option?: { keepEmpty?: boolean },
): ReactElement[] => {
	const keepEmpty = option && option.keepEmpty;
	let ret: React.ReactElement[] = [];

	Children.forEach(children, (child: any) => {
		if ((child === undefined || child === null) && !keepEmpty) {
			return;
		}

		if (Array.isArray(child)) {
			ret = ret.concat(toElementArray(child));
		} else if (isFragment(child) && child.props) {
			ret = ret.concat(toElementArray(child.props.children, option));
		} else {
			ret.push(child);
		}
	});

	return ret;
};

export const createForwardRef = <R, P = any>(
	displayName: string,
	render: ForwardRefRenderFunction<R, P>,
): ForwardRefExoticComponent<PropsWithoutRef<P> & RefAttributes<R>> => {
	const fc = forwardRef(render);
	fc.displayName = displayName;
	return fc;
};

export const createDefaultsForwardRef = <R, P, D>(
	displayName: string,
	defaultProps: D,
	render: ForwardRefRenderFunction<R, P & D>,
): ForwardRefExoticComponent<
	PropsWithoutRef<P & Partial<D>> & RefAttributes<R>
> => {
	const fc = forwardRef(render);
	fc.displayName = displayName;
	fc.defaultProps = defaultProps;
	return fc as any;
};

export const createFC = <P = any>(
	displayName: string,
	fc: FunctionComponent<P>,
): FunctionComponent<P> => {
	fc.displayName = displayName;
	return fc;
};

export const createDefaultsFC = <P, D>(
	displayName: string,
	defaultProps: D,
	fc: FunctionComponent<P & D>,
): FunctionComponent<P & Partial<D>> => {
	fc.displayName = displayName;
	fc.defaultProps = defaultProps;
	return fc as any;
};
