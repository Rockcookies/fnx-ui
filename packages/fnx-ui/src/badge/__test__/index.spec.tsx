import React from 'react';
import { render } from '@testing-library/react';
import Badge from '../index';

describe('<Badge/>', () => {
	it('should render children correctly', () => {
		const { container } = render(<Badge>children</Badge>);

		expect(container).toMatchSnapshot();
	});

	it('should render default count when children is not exist', () => {
		const { container } = render(<Badge></Badge>);

		expect(container).toMatchSnapshot();
	});

	it('should change count position when using offset prop', () => {
		const { container, rerender } = render(
			<Badge offset={[10, -20]} count={8}>
				children
			</Badge>,
		);

		expect(container).toMatchSnapshot();

		rerender(
			<Badge offset={['-10', '20']} count={8}>
				children
			</Badge>,
		);

		expect(container).toMatchSnapshot();
	});

	it('should change dot position when using offset prop with custom unit', () => {
		const { container } = render(
			<Badge offset={['1rem', '2rem']} dot>
				children
			</Badge>,
		);

		expect(container).toMatchSnapshot();
	});

	it('should change dot position when using offset prop without children', () => {
		const { container } = render(<Badge offset={[10, 20]} dot></Badge>);

		expect(container).toMatchSnapshot();
	});

	it('should render nothing when count is zero', () => {
		const { container } = render(<Badge count={0}>children</Badge>);

		expect(container).toMatchSnapshot();
	});

	it('should not render zero when show-zero is false', () => {
		const { container } = render(
			<Badge showZero={false} count={0}>
				children
			</Badge>,
		);

		expect(container).toMatchSnapshot();
	});
});
