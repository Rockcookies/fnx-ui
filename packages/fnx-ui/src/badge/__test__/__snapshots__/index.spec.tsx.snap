// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Badge/> should change count position when using offset prop 1`] = `
<div>
  <div
    class="fnx-badge__wrapper"
  >
    children
    <span
      class="fnx-badge fnx-badge--top-right fnx-badge--fixed"
      style="top: 20px; right: -10px;"
    >
      8
    </span>
  </div>
</div>
`;

exports[`<Badge/> should change count position when using offset prop 2`] = `
<div>
  <div
    class="fnx-badge__wrapper"
  >
    children
    <span
      class="fnx-badge fnx-badge--top-right fnx-badge--fixed"
      style="top: -20px; right: 10px;"
    >
      8
    </span>
  </div>
</div>
`;

exports[`<Badge/> should change dot position when using offset prop with custom unit 1`] = `
<div>
  <div
    class="fnx-badge__wrapper"
  >
    children
    <span
      class="fnx-badge fnx-badge--top-right fnx-badge--dot fnx-badge--fixed"
    />
  </div>
</div>
`;

exports[`<Badge/> should change dot position when using offset prop without children 1`] = `
<div>
  <span
    class="fnx-badge fnx-badge--top-right fnx-badge--dot"
    style="margin-top: 20px; margin-left: 10px;"
  />
</div>
`;

exports[`<Badge/> should not render zero when show-zero is false 1`] = `
<div>
  <div
    class="fnx-badge__wrapper"
  >
    children
    <span
      class="fnx-badge fnx-badge--top-right fnx-badge--fixed"
    />
  </div>
</div>
`;

exports[`<Badge/> should render children correctly 1`] = `
<div>
  <div
    class="fnx-badge__wrapper"
  >
    children
    <span
      class="fnx-badge fnx-badge--top-right fnx-badge--fixed"
    >
      8
    </span>
  </div>
</div>
`;

exports[`<Badge/> should render default count when children is not exist 1`] = `
<div>
  <span
    class="fnx-badge fnx-badge--top-right"
  >
    8
  </span>
</div>
`;

exports[`<Badge/> should render nothing when count is zero 1`] = `
<div>
  <div
    class="fnx-badge__wrapper"
  >
    children
    <span
      class="fnx-badge fnx-badge--top-right fnx-badge--fixed"
    >
      0
    </span>
  </div>
</div>
`;
